src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp

```diff
@@ -902,23 +911,38 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalStartBackgroundWork(_In_ BackgroundCall
         ResumeThread(hThread);
     }
 
+    if (name && g_pfnSetThreadDescription)
+    {
+        size_t len = strlen(name);
+        wchar_t* threadNameWide = new (nothrow) wchar_t[len + 1];
+        if (threadNameWide != nullptr)
+        {
+            if (!MultiByteToWideChar(CP_UTF8, 0, name, -1, threadNameWide, (int)(len + 1)) == 0)
+            {
+                threadNameWide[len] = '\0';
+                g_pfnSetThreadDescription(hThread, threadNameWide);
+            }
+            delete[] threadNameWide;
+        }
+    }
+
     CloseHandle(hThread);
     return true;
 }
 
-REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalStartBackgroundGCThread(_In_ BackgroundCallback callback, _In_opt_ void* pCallbackContext)
+REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalStartBackgroundGCThread(_In_ BackgroundCallback callback, _In_opt_ void* pCallbackContext, const char* name)
 {
-    return PalStartBackgroundWork(callback, pCallbackContext, FALSE);
+    return PalStartBackgroundWork(callback, pCallbackContext, FALSE, name);
 }
 
 REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalStartFinalizerThread(_In_ BackgroundCallback callback, _In_opt_ void* pCallbackContext)
 {
-    return PalStartBackgroundWork(callback, pCallbackContext, TRUE);
+    return PalStartBackgroundWork(callback, pCallbackContext, TRUE, nullptr);
```

Comment:
Should we set the thread names for other threads too here and also in `ThreadNameChanged` for managed threads?

Reply:
I added the names for the finalizer thread and the event pipe server thread. I am not sure about the `ThreadNameChanged` scenario.

Reply:
>  I am not sure about the ThreadNameChanged scenario.

Uncomment [`ThreadNameChanged`](https://github.com/dotnet/runtime/blob/a16e0d58e92fe5f3bcda9e75e9afd1ee0f766967/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.cs#L201-L202) method and make it call SetThreadName that you have added.

Reply:
Right now, my scenario can be satisfied by only setting a thread's own name. Setting the name of some other threads are complicated (e.g. on Linux, we need to get the `pthread_t` instance) or impossible (e.g. on OSX, this is just impossible). I think we should scope that out of the current PR.

Reply:
Sounds good

Reply:
I think it is fine to scope this to setting name of current thread only. But I think it would be nice to hook up `ThreadNameChanged` for the current thread. It should be ~10 lines.